crates/sui-indexer-alt-jsonrpc/src/api/coin.rs

```diff
@@ -0,0 +1,349 @@
+    }
+
+    async fn get_all_coins(
+        &self,
+        owner: SuiAddress,
+        cursor: Option<String>,
+        limit: Option<usize>,
+    ) -> RpcResult<JsonRpcPage<Coin, String>> {
+        Ok(self
+            .get_coins_impl(owner, None, cursor, limit)
+            .await
+            .with_internal_context(|| format!("Failed to get coins for {owner}"))?)
+    }
+}
+
+impl Coins {
+    async fn get_coins_impl(
+        &self,
+        owner: SuiAddress,
+        coin_type_tag: Option<TypeTag>,
+        cursor: Option<String>,
+        limit: Option<usize>,
+    ) -> Result<JsonRpcPage<Coin, String>, RpcError<Error>> {
+        let Self(_, config) = self;
+
+        let page: Page<BalanceCursor> = Page::from_params(
+            config.default_page_size,
+            config.max_page_size,
+            cursor,
+            limit,
+            None,
+        )?;
+
+        // We get all the qualified coin ids first.
+        let coin_id_page = self.get_coin_id_page(owner, coin_type_tag, page).await?;
+
+        let coin_futures = coin_id_page.data.iter().map(|id| self.coin_response(*id));
+
+        let coins = future::join_all(coin_futures)
+            .await
+            .into_iter()
+            .zip(coin_id_page.data)
+            .map(|(r, c)| {
+                let id = c;
+                r.with_internal_context(|| format!("Failed to get object {id}"))
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        Ok(JsonRpcPage {
+            data: coins,
+            next_cursor: coin_id_page.next_cursor,
+            has_next_page: coin_id_page.has_next_page,
+        })
+    }
+
+    async fn get_coin_id_page(
+        &self,
+        owner: SuiAddress,
+        coin_type_tag: Option<TypeTag>,
+        page: Page<BalanceCursor>,
+    ) -> Result<JsonRpcPage<ObjectID, String>, RpcError<Error>> {
+        use coin_balance_buckets::dsl as cb;
+
+        let Self(ctx, _) = self;
+        let mut conn = ctx
+            .reader()
+            .connect()
+            .await
+            .context("Failed to connect to database")?;
+
+        let limit = page.limit;
+
+        // We use two aliases of coin_balance_buckets to make the query more readable.
+        let (candidates, newer) = diesel::alias!(
+            coin_balance_buckets as candidates,
+            coin_balance_buckets as newer
+        );
+
+        // Macros to make the query more readable.
+        macro_rules! candidates {
+            ($($field:ident),*) => {
+                candidates.fields(($(cb::$field),*))
+            }
+        }
+
+        macro_rules! newer {
+            ($($field:ident),*) => {
+                newer.fields(($(cb::$field),*))
+            }
+        }
+
+        // Construct the basic query first to filter by owner, not deleted and newest rows.
+        let mut query = candidates
+            .select(candidates!(
+                object_id,
+                cp_sequence_number,
+                coin_balance_bucket
```

Comment:
Did you try adding `assume_not_null` here? (based on the discussion below around `IdCpBalance` needing to use an optional field for this?)